### PR TITLE
[Communication] - Identifier serialization new schema

### PR DESCRIPTION
**Pull request by RezaJooyandeh**

### CODE REVIEW

File: sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs [hunk 1]
```diff
@@ -8,28 +8,38 @@ namespace Azure.Communication
     /// <summary>Represents a Microsoft Teams user.</summary>
     public class MicrosoftTeamsUserIdentifier : CommunicationIdentifier
     {
+        /// <summary>The id of the Microsoft Teams User identifier.</summary>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs:11]: Let's use the word `full` to distinguish Id better from UserId

```suggestion
        /// <summary>The full id of the Microsoft Teams User identifier.</summary>
```

File: sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs [hunk 2]
```diff
@@ -8,28 +8,38 @@ namespace Azure.Communication
...
+        public string? Id { get; }
+
         /// <summary>The id of the Microsoft Teams user. If the user isn't anonymous, the id is the AAD object id of the user.</summary>
         public string UserId { get; }
 
         /// <summary>True if the user is anonymous, for example when joining a meeting with a share link.</summary>
         public bool IsAnonymous { get; }
 
+        /// <summary> The cloud that the identifier belongs to. </summary>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs:20]: nit: extra space

```suggestion
        /// <summary>The cloud that the identifier belongs to. </summary>
```

File: sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs [hunk 3]
```diff
@@ -8,28 +8,38 @@ namespace Azure.Communication
...
+        public CommunicationCloudEnvironment Cloud { get; }
+
         /// <summary>
         /// Initializes a new instance of <see cref="MicrosoftTeamsUserIdentifier"/>.
         /// </summary>
         /// <param name="userId">Id of the Microsoft Teams user. If the user isn't anonymous, the id is the AAD object id of the user.</param>
         /// <param name="isAnonymous">Set this to true if the user is anonymous, for example when joining a meeting with a share link.</param>
+        /// <param name="id">Id of the Microsoft Teams user.</param>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs:28]:
```suggestion
        /// <param name="id">Full id of the Microsoft Teams user, optional.</param>
```

File: sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs [hunk 4]
```diff
@@ -8,28 +8,38 @@ namespace Azure.Communication
...
+        /// <param name="cloud">Id of the Microsoft Teams user.</param>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs:29]:
```suggestion
        /// <param name="cloud">The cloud that the Microsoft Team user belongs to. A null value translates to the Public cloud.</param>
```

File: sdk/communication/Azure.Communication.Common/src/PhoneNumberIdentifier.cs [hunk 5]
```diff
@@ -9,21 +9,26 @@ namespace Azure.Communication
...
     {
+        /// <summary>The id of the phone number.</summary>
+        public string? Id { get; }
+
         /// <summary>The phone number in E.164 format.</summary>
         public string PhoneNumber { get; }
 
         /// <summary> Initializes a new instance of <see cref="PhoneNumberIdentifier"/>.</summary>
         /// <param name="phoneNumber">The phone number in E.164 format.</param>
+        /// <param name="id">Id of the Phone Number.</param>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/PhoneNumberIdentifier.cs:20]: nit: casing and let's use "full"
```suggestion
        /// <param name="id">Full id of the phone number.</param>
```

File: sdk/communication/Azure.Communication.Common/src/PhoneNumberIdentifier.cs [hunk 6]
```diff
@@ -9,21 +9,26 @@ namespace Azure.Communication
     /// <summary>Represents a Phone Number.</summary>
     public class PhoneNumberIdentifier : CommunicationIdentifier
     {
+        /// <summary>The id of the phone number.</summary>
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/src/PhoneNumberIdentifier.cs:12]:
```suggestion
        /// <summary>The full id of the phone number.</summary>
```

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs [hunk 7]
```diff
@@ -98,73 +99,82 @@ public void SerializeCallingApplication()
...
+        public void SerializeMicrosoftTeamsUser(bool isAnonymous, string? expectedId)
         {
-            CommunicationIdentifierModel model = CommunicationIdentifierSerializer.Serialize(new MicrosoftTeamsUserIdentifier("some id", isAnonymous));
+            CommunicationIdentifierModel model = CommunicationIdentifierSerializer.Serialize(new MicrosoftTeamsUserIdentifier("user id", isAnonymous, expectedId));
 
             Assert.AreEqual(CommunicationIdentifierKind.MicrosoftTeamsUser, model.Kind);
-            Assert.AreEqual("some id", model.MicrosoftTeamsUserId);
+            Assert.AreEqual("user id", model.MicrosoftTeamsUserId);
             Assert.AreEqual(isAnonymous, model.IsAnonymous);
+            Assert.AreEqual(expectedId, model.Id);
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:155]: Let's add an Assert that Cloud is set to Public when it's missing. Alos that IsAnonymous is set to false when it's missing in the constructor call.

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs [hunk 8]
```diff
@@ -98,73 +99,82 @@ public void SerializeCallingApplication()
...
         }
 
         [Test]
         [TestCase(false)]
         [TestCase(true)]
         public void DeserializerMicrosoftTeamsUser(bool isAnonymous)
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:161]:
```suggestion
        public void DeserializeMicrosoftTeamsUser(bool isAnonymous)
```

File: sdk/communication/Shared/src/CommunicationIdentifierSerializer.cs [hunk 9]
```diff
@@ -18,28 +18,73 @@ public static CommunicationIdentifier Deserialize(CommunicationIdentifierModel i
...
+        private static CommunicationCloudEnvironment Deserialize(CommunicationCloudEnvironmentModel cloud)
+        {
+            if (cloud == CommunicationCloudEnvironmentModel.Public)
+                return CommunicationCloudEnvironment.Public;
+            if (cloud == CommunicationCloudEnvironmentModel.Gcch)
+                return CommunicationCloudEnvironment.Gcch;
+            if (cloud == CommunicationCloudEnvironmentModel.Dod)
+                return CommunicationCloudEnvironment.Dod;
+
+            throw new NotSupportedException();
```
Comment (DominikMe) [inline sdk/communication/Shared/src/CommunicationIdentifierSerializer.cs:43]: That enum doesn't seem very extensible to me, when we throw here...

Shouldn't we return `new CommunicationCloudEnvironment(CommunicationCloudEnvironmentModel.ToString())`?

File: sdk/communication/Shared/src/CommunicationIdentifierSerializer.cs [hunk 10]
```diff
@@ -18,28 +18,73 @@ public static CommunicationIdentifier Deserialize(CommunicationIdentifierModel i
...
+        private static CommunicationCloudEnvironmentModel Serialize(CommunicationCloudEnvironment cloud)
+        {
+            if (cloud == CommunicationCloudEnvironment.Public)
+                return CommunicationCloudEnvironmentModel.Public;
+            if (cloud == CommunicationCloudEnvironment.Gcch)
+                return CommunicationCloudEnvironmentModel.Gcch;
+            if (cloud == CommunicationCloudEnvironment.Dod)
+                return CommunicationCloudEnvironmentModel.Dod;
+
+            throw new NotSupportedException();
```
Comment (DominikMe) [inline sdk/communication/Shared/src/CommunicationIdentifierSerializer.cs:85]: same comment, pass the string through instead?

`new CCEM(cce.ToString())`

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs [hunk 11]
```diff
@@ -98,73 +99,82 @@ public void SerializeCallingApplication()
...
-            Assert.True(identifier is MicrosoftTeamsUserIdentifier);
-            Assert.AreEqual(expectedIdentifier.UserId, ((MicrosoftTeamsUserIdentifier)identifier).UserId);
-            Assert.AreEqual(expectedIdentifier.IsAnonymous, ((MicrosoftTeamsUserIdentifier)identifier).IsAnonymous);
+            Assert.AreEqual(expectedIdentifier.UserId, identifier.UserId);
+            Assert.AreEqual(expectedIdentifier.IsAnonymous, identifier.IsAnonymous);
+            Assert.AreEqual(expectedIdentifier.Cloud, identifier.Cloud);
+            Assert.AreEqual(expectedIdentifier.Id, identifier.Id);
             Assert.AreEqual(expectedIdentifier, identifier);
         }
     }
```
Comment (DominikMe) [inline sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:180]: Do we have a case that tests that we throw when `Cloud` is missing during deserialization?

File: sdk/communication/Shared/src/CommunicationIdentifierModel.cs [hunk 12]
```diff
@@ -27,5 +27,7 @@ internal CommunicationIdentifierModel(CommunicationIdentifierKind kind)
         public string MicrosoftTeamsUserId { get; set; }
         /// <summary> True if the identifier is anonymous. </summary>
         public bool? IsAnonymous { get; set; }
+        /// <summary> The cloud that the identifier belongs to. </summary>
+        public CommunicationCloudEnvironmentModel? Cloud { get; set; }
```
Comment (DominikMe) [inline sdk/communication/Shared/src/CommunicationIdentifierModel.cs:31]: Why are `Cloud` and `IsAnonymous` nullable but `PhoneNumber` and `MicrosoftTeamsUserId` not?

**Review by DominikMe — COMMENTED** [review]
Looks mostly good to me, some comments.

File: sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs (hunk 2 above)
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Common/src/MicrosoftTeamsUserIdentifier.cs:20]: That is a autogen issue. But does not affect the generated doc. It is ok.

File: sdk/communication/Shared/src/CommunicationIdentifierModel.cs (hunk 12 above)
Comment (RezaJooyandeh) [reply sdk/communication/Shared/src/CommunicationIdentifierModel.cs:31]: Autorest treats ref-types as nullable and is not using non-nullable ref-types

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs (hunk 7 above)
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:155]: I have a test in the identifier types for this as in C# we do it during construction.

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs (hunk 11 above)
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:180]: yes

**Review by DominikMe — APPROVED** [review]

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs [hunk 13]
```diff
@@ -17,8 +17,11 @@ public void MissingProperty_DeserializerThrows()
...
-                new CommunicationIdentifierModel(CommunicationIdentifierKind.MicrosoftTeamsUser) { Id = "some id", MicrosoftTeamsUserId = "some id" }, // Missing IsAnonymous
-                new CommunicationIdentifierModel(CommunicationIdentifierKind.MicrosoftTeamsUser) { Id = "some id", IsAnonymous = true }, // Missing MicrosoftTeamsUserId
+                new CommunicationIdentifierModel(CommunicationIdentifierKind.PhoneNumber) { PhoneNumber = "+12223334444" }, // Missing Id
+                new CommunicationIdentifierModel(CommunicationIdentifierKind.MicrosoftTeamsUser) { Id = "some id", MicrosoftTeamsUserId = "some id", Cloud = CommunicationCloudEnvironmentModel.Public }, // Missing IsAnonymous
+                new CommunicationIdentifierModel(CommunicationIdentifierKind.MicrosoftTeamsUser) { Id = "some id", IsAnonymous = true, Cloud = CommunicationCloudEnvironmentModel.Public }, // Missing MicrosoftTeamsUserId
+                new CommunicationIdentifierModel(CommunicationIdentifierKind.MicrosoftTeamsUser) { MicrosoftTeamsUserId = "some id", IsAnonymous = true, Cloud = CommunicationCloudEnvironmentModel.Public }, // Missing Id
```
Comment (JianpingChen) [inline sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:23]: But in the constructor of MicrosoftTeamsUserIdentifier, id is optional. Should we throw in this case?

**Review by JianpingChen — CHANGES_REQUESTED** [review]

File: sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs (hunk 13 above)
Comment (RezaJooyandeh) [reply sdk/communication/Azure.Communication.Common/tests/CommunicationIdentifierSerializerTest.cs:23]: Even though the id for MicrosoftTeamsUserIdenfier is optional to be set by the developers, it is mandatory to be sent by the services. Hence deserialization assumes that it is always present

**Review by JianpingChen — APPROVED** [review]

**Pull request merged**